test/integration/puppetlabs/general_puppet/general_puppet_int_test.clj

```diff
@@ -68,11 +70,32 @@
       app {:jruby-puppet
            {:max-active-instances num-jrubies}
            :versioned-code
-           {:code-id-command (script-path "warn_echo_and_error")}}
+           {:code-id-command (script-path "warn_echo_and_error")
+            :code-content-command (script-path "echo")}}
       (let [catalog (testutils/get-catalog)]
         (is (nil? (get catalog "code_id")))
         (is (logged? #"Non-zero exit code returned while running" :error))
-        (is (logged? #"Executed an external process which logged to STDERR: production" :warn)))))))
+        (is (logged? #"Executed an external process which logged to STDERR: production" :warn))))))
+  (testing "code id is not added and 500 is returned if code-content-command is not set"
+    (logging/with-test-logging
+     (bootstrap/with-puppetserver-running
+      app {:jruby-puppet
+           {:max-active-instances num-jrubies}
+           :versioned-code
+           {:code-id-command (script-path "echo")}}
+      (let [response (testutils/http-get "puppet/v3/catalog/localhost?environment=production")]
+        (is (= 500 (:status response)))
+        (is (logged? #"Error 500 on SERVER"))))))
```

Comment:
The discussion in the meeting the other day mentioned that for any catalog compilation that is not using Direct Puppet, the catalog compilation should still succeed.

IIRC, the determination of whether we're using DP or not will be based on a combination of:
1. did the agent send up a query param indicating that they support static catalogs?
2. is the per-environment setting for static catalogs enabled for this environment?

The first question should be easy to answer.  The second will not be, because it would require us to dig into the environment.conf file, which is complex and something we won't be able to do for this release.  Perhaps we need to discuss this again in the DP meeting today; it may be a situation where we just need to send some value along to the Ruby code and let them make the determination of whether or not the catalog needs to be failed.

Either that, or we just fail to start up if the server is misconfigured.  That'd obviously be much easier so perhaps we should ask at the meeting whether there are any significant downsides to that.